src/compiler/checker.ts

```diff
@@ -19081,14 +19098,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (checkType === wildcardType || extendsType === wildcardType) {
                 return wildcardType;
             }
+            const effectiveCheckType = forNarrowing && isNarrowingSubstitutionType(checkType)

```

Comment:
Why do you need both `forNarrowing` and to check the flags of the check type?

Reply:
`forNarrowing = true` is passed by `getConditionalTypeInstantiation` when we validate the conditional type is safe to narrow. It could be that the check type is a narrowing substitution type, but we don't want to narrow the conditional type because it has an invalid shape.